### ISSUE TITLE
deno watch for web development

### ISSUE THREAD

**Issue by MarkTiedemann** [denoland/deno#2401]
Would love to have a better web development experience with Deno out of the box.

- First, the server should restart whenever I make changes.
- Second, the website should reload whenever I make changes.

For the server, I'd like to propose a `deno watch` subcommand or `deno run --watch` flag that automatically restarts the server in a controlled way, e.g.

```
deno watch *.ts --run main.ts --allow-net
# or
deno run --watch *.ts --allow-net main.ts
```

If we also had a `Deno.watch` API for watching directories or files, we could then implement a simple client refresh via [server-sent events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events) ([example](https://gist.github.com/MarkTiedemann/1c43c42792825d2a86f88bbd52f55a5a)) or web sockets.

EDIT: It would be nice if the `watch` subcommand/flag would be able to notify the running code that it will/should be restarted - maybe something via signals (e.g. `SIGUSR1`), command line flags, some API, or a similar mechanism - so that the server could notify the client that it will be restarted and that the client needs to be reloaded.

**Comment by kitsonk** [comment]
Duplicate of #1826

**Comment by MarkTiedemann** [comment]
> Duplicate of #1826

I'd say it's related rather than a duplicate. #1826 is about the `Deno.watch` API, not about the `watch` subcommand or `--watch` flag.

I also found #2251 which is related. It's about "automatically restarting the server". I think the solution to that would be the subcommand / flag.

**Comment by ry** [comment]
Once #1826 is done, I would consider this. Tho it would need to be carefully specified - and I'm not sure if it can be accomplished in general.

**Comment by Caesar2011** [comment]
Now, when #1826 is done, we can consider this again. Based on another issue I implemented a small typescript script.

https://github.com/Caesar2011/rhinoder

This could at least be used for `Deno.watch()`.

**Comment by eliassjogreen** [comment]
I have made [denon](https://github.com/eliassjogreen/denon) which serves the same purpose but with a bit more configurability.

**Comment by MarkTiedemann** [comment]
Don't think that this should be part of Deno core anymore, closing.

**Comment by rudyhuynh** [comment]
@MarkTiedemann I think this should be part of Deno core. We already have unit test, code format, bundler, .etc. Watch for development must be major piece of Deno core as well.

**Comment by Caesar2011** [comment]
Consider `deno run --watch file.ts`.

**Comment by MarkTiedemann** [comment]
First of all, I think Deno is fast enough that you don't really need a watcher. Just `Ctrl+C`, `Up`, `Enter` is probably fine for 90% of use cases. It has been working quite fine for me working with Deno on a daily basis for the past one and a half years. You should try it. :)

Then there's two arguments I made when originally opening this issue:

1) The server should restart whenever I make changes.
2) The website should reload whenever I make changes.

2, I think, is out of scope anyways. Deno should not have live-reload integration or some hacky server-sent events that trigger in-browser refresh, as I originally proposed. In retrospect, I think that is a dumb idea. If you really need something like that, write it yourself, put it on GitHub - but I don't think this should ever be in core.

1, I think, is doable and could potentially be in core some day. But then again, I don't think `deno run --watch app.ts` is any better than `denon run app.ts`, or using the usual general purpose filewatching tools that have long solved this problem (https://github.com/emcrisostomo/fswatch, https://github.com/facebook/watchman, etc.)

If someone wants to open another issue for this, feel free. This one I'd like to keep closed.

**Comment by jtwebman** [comment]
@MarkTiedemann Closing the whole issue because 2 isn't a good idea but 1 would be a big win seems very dictator-ish. It is very common when working on an API application to restart deno. The sad part is right now I use Node.js nodemon to do it.

`npx nodemon --exec "deno run --allow-net=0.0.0.0" server.js`

But will gladly open another issue to be more clear with what is needed and the win for having it the core. `--watch` in the run command will allow for you to only watch the files that are loaded with the command vs restarting if anything changes.

**Comment by Qu4k** [comment]
Also, there is [`denosaurs/denon`](https://github.com/denosaurs/denon)

**Comment by rwieruch** [comment]
I was surprised seeing that there is no watch mode for development purposes in Deno (correct me if I am wrong). After going through Deno's philosophy, IMO it shouldn't be the case that I already have to fallback to nodemon or a third-party (?) like denon for this. This is a core developer experience issue for me.

**Comment by pedropaulosuzuki** [comment]
Just in case someone is wondering, #7382 got merged successfully 22 days ago

**Comment by tbsvttr** [comment]
`deno bundle --unstable --watch ./src/app.ts ./build/app.bundle.ts` works for me with deno 1.6.0.

**Comment by asbjornu** [comment]
Seems like `--watch` supports globbing and multiple paths separated by comma as well. This works for me, at least:

```shell
deno run --watch=./**/*.ts,./**/*.tsx server.ts
```

**Comment by bebraw** [comment]
@asbjornu Through research at https://github.com/jurassiscripts/velociraptor/issues/80 we found out `watch` supports only a boolean flag right now. I opened #12197 with a small proposal.
